saffron/src/lib.rs

```diff
@@ -1359,194 +1360,128 @@ impl Cron {
+                Some(next_month) => next_month,
+                None => return Ok(None),
+            };
+
+            date = match self.find_next_month(date) {
+                Some(date) if date > bound => return Err(OutOfBound),
+                Some(date) => date,
+                None => return Ok(None),
+            };
 
-                return Some(NaiveTime::from_hms(next_hour, minute, 0));
+            match self.find_next_day(date) {
+                Some(next_day) if next_day > bound => return Err(OutOfBound),
+                Some(next_day) => return Ok(Some(next_day)),
+                None => {}
             }
         }
-
-        None
     }
+}
 
-    fn find_next_date(&self, date: Date<Utc>) -> Option<Date<Utc>> {
-        const VALID_NEXT_DAY: &str = "Day should be valid for giving month";
-        const VALID_NEXT_MONTH: &str = "Month should be valid";
+struct OutOfBound;
 
-        let mut month = date.month();
-        if self.months.contains_month(date) {
-            if let Some(next_day) = self.find_next_day(date) {
-                return Some(date.with_day0(next_day).expect(VALID_NEXT_DAY));
-            }
-        }
+#[inline]
+fn minute_floor(dt: DateTime<Utc>) -> DateTime<Utc> {
+    dt.with_second(0)
+        .expect("zero is a valid second value")
+        .with_nanosecond(0)
+        .expect("zero is a valid nanosecond value")
+}
 
-        while month <= 11 {
-            if let Some(next_month) = self.find_next_month(month) {
-                month = next_month;
-
-                let start_month_date = date
-                    .with_day(1)
-                    .expect(VALID_NEXT_DAY)
-                    .with_month0(month)
-                    .expect(VALID_NEXT_MONTH);
-                if let Some(month_day) = self.find_next_day(start_month_date) {
-                    return Some(start_month_date.with_day0(month_day).expect(VALID_NEXT_DAY));
-                } else {
-                    month += 1;
-                }
-            } else {
-                break;
-            }
-        }
+#[inline]
+fn previous_minute(dt: DateTime<Utc>) -> Option<DateTime<Utc>> {
+    dt.checked_sub_signed(Duration::minutes(1))
+}
+
+#[inline]
+fn next_minute(dt: DateTime<Utc>) -> Option<DateTime<Utc>> {
+    dt.checked_add_signed(Duration::minutes(1))
+}
 
+/// Gets the next month in the year if one exists.
+#[inline]
+fn next_month_in_year(d: Date<Utc>) -> Option<Date<Utc>> {
+    let month = d.month();
+    if month <= 11 {
+        Utc.ymd_opt(d.year(), month + 1, 1).single()
+    } else {
         None
     }
 }
 
-enum IterTime {
-    /// Return the current time if it matches, otherwise change to `Next` and get the next time
-    Start(DateTime<Utc>),
-    /// Return the next time that matches the cron expression after this one
-    Next(DateTime<Utc>),
-    /// The end of the iterator
-    End,
+#[inline]
+fn maybe_time_bound(d: Date<Utc>, bound: DateTime<Utc>) -> Option<NaiveTime> {
```

Comment:
This function's name is kinda confusing. It's checking if `d` is the date of `bound`. If so, return the bound's time, if not, return None. I'm having a hard time thinking of a good name based on the function itself (`timeBoundOfDate`? ) Maybe a good name would be hinting towards why you'd want to call this function. It's definitely not clear to me what this function is useful for.